src/dsl/compute_graph.jl

```diff
@@ -0,0 +1,180 @@
+    !isdefined(parent, :left) || error("left child is already assigned")
+    node = typeof(parent)(data, parent)
+    parent.left = node
+end
+
+"""
+    rightchild(data, parent::BinaryNode)
+
+Returns right child node or nothing
+"""
+function rightchild(data, parent::BinaryNode)
+    !isdefined(parent, :right) || error("right child is already assigned")
+    node = typeof(parent)(data, parent)
+    parent.right = node
+end
+
+"""
+    AbstractTrees.children(node::BinaryNode)
+
+Implement children function from AbstractTrees package
+"""
+function AbstractTrees.children(node::BinaryNode)
+    if isdefined(node, :left)
+        if isdefined(node, :right)
+            return (node.left, node.right)
+        end
+        return (node.left,)
+    end
+    isdefined(node, :right) && return (node.right,)
+    return ()
+end
+
+## functions and data structures for trees of contraction commands
+
+"""Structure to represent a contraction command"""
+struct ContractCommand
+    output_name::Symbol
+    output_idxs::Vector{Int}
+    left_name::Symbol
+    left_idxs::Vector{Int}
+    right_name::Symbol
+    right_idxs::Vector{Int}
+end
+
+"""
+    Base.write(io::IO, cmd::ContractCommand)
+
+Function to serialise command to the given IO stream
+"""
+function Base.write(io::IO, cmd::ContractCommand)
+    j = x -> length(x) == 0 ? "0" : join(x, ",")
+    write(io, "ncon $(cmd.output_name) $(j(cmd.output_idxs)) $(cmd.left_name) $(j(cmd.left_idxs)) $(cmd.right_name) $(j(cmd.right_idxs))\n")
+end
+
+"""
+    Base.write(io::IO, tree::tree::BinaryNode{ContractCommand})
+
+Function to serialise tree to the given IO stream
+"""
+function Base.write(io::IO, tree::BinaryNode{ContractCommand})
+    for cmd in PostOrderDFS(tree)
+        write(io, cmd.data)
+    end
+end
+
+"""
+    build_tree(cmds::Vector{ContractCommand})
+
+Function to construct a tree from a list of contraction commands
+"""
+function build_tree(cmds::Vector{ContractCommand})
+    cmd_map = Dict(x.output_name => deepcopy(x) for x in cmds)
+    all_outputs = Set(keys(cmd_map))
+    get_names = x -> [x.left_name, x.right_name]
+    all_inputs = vcat(get_names.(values(cmd_map))...)
+    root_sym = collect(setdiff(all_outputs, all_inputs))[1]
+    root_node = _create_node(cmd_map, root_sym)
+    root_node
+end
+
+function _create_node(cmd_map, cmd_sym)
+    cmd = cmd_map[cmd_sym]
+    node = BinaryNode{ContractCommand}(cmd)
+    if haskey(cmd_map, cmd.left_name)
+        node.left = _create_node(cmd_map, cmd.left_name)
+        node.left.parent = node
+    end
+    if haskey(cmd_map, cmd.right_name)
+        node.right = _create_node(cmd_map, cmd.right_name)
+        node.right.parent = node
+    end
+    node
+end
+
+AbstractTrees.printnode(io::IO, node::BinaryNode{ContractCommand}) = print(io, node.data.output_name)
+
+"""
+    remove_repeated!(indices::Vector{Int64})
+
+Given a list of integers, only keep first occurence of each
+"""
+function remove_repeated!(indices::Vector{Int64})
```

Comment:
Can `Base.unique!` be used for this?

Reply:
Yes, I don't see why not. Not sure why I didn't consider this.